### DISCUSSION TITLE
Kalman gain

### DISCUSSION THREAD

**Discussion by Mtulaza1** [rodralez/NaveGo#153] [Q&A] [answered]
Dear NaveGo team,
I have struggle with tuning and stability of Kalman filter, I would like to improve Kalman gain for my implemented set of data. It is described by P.D. Grooves in section 3.3.1 that "The tuning of the Kalman filter is the selection by the designer or user of values
for three matrices. These are the system noise covariance matrix, Qk, the measurement noise covariance matrix, Rk, and the initial values of the error covariancematrix, P+0. It is important to select these parameters correctly.". I also read the followed article where mathematical model of NaveGo is presented but I have no idea how to corectly change those parameters in your code to obtain better Kalman gain.
Thank for your respond.

**Comment by rodralez** [comment, marked as answer]
Hello,

I will share my experience in tuning the Kalman filter.

1. Q, the system noise covariance matrix, sets how much you trust your model. Smaller values, you trust your model pretty much.
2. R, the measurement noise covariance matrix, sets the quality of the system output sensors as GNSS. Values come from sensor manufacturers' datasheets.
3. P, the error covariance matrix, sets the initial errors when the system starts. Choose bigger errors since the KF will rapidly converge to smaller errors.

So, start with P and Q with bigger values, and R with manufacturers' values. 

If your system doesn't diverge you are good. If not, typically increase Q.

Once your system doesn't diverge, try to decrease R in baby steps. If everything goes fine, try to decrease Q in baby steps.

Be methodical, just modify a matrix, or even one value at matrix diagonal, at a time. 

Regards.
